commcare-cloud-bootstrap/commcare_cloud_bootstrap.py

```diff
@@ -189,15 +299,72 @@ def copy_default_vars(env, aws_config):
               file=sys.stderr)
 
 
+class Provision(object):
+    command = 'provision'
+    help = """Provision a new environment based on a spec yaml file. (See example_spec.yml.)"""
+
+    @staticmethod
+    def make_parser(parser):
+        parser.add_argument('spec')
+        parser.add_argument('--env')
+
+    @staticmethod
+    def run(args):
+        with open(args.spec) as f:
+            spec = yaml.load(f)
+
+        spec = Spec.wrap(spec)
+        provision_machines(spec, args.env)
+
+
+class Show(object):
+    command = 'show'
+    help = """Show provisioned instances for a given env"""
+
+    @staticmethod
+    def make_parser(parser):
+        parser.add_argument('env')
+
+    @staticmethod
+    def run(args):
+        describe_instances = raw_describe_instances(args.env)
+        print_describe_instances(describe_instances)
+
+
+class Reip(object):
+    command = 'reip'
+    help = ("Rewrite the public IP addresses in the inventory for an env. "
+            "Useful after reboot.")
+
+    @staticmethod
+    def make_parser(parser):
+        parser.add_argument('env')
+
+    @staticmethod
+    def run(args):
+        describe_instances = raw_describe_instances(args.env)
+        new_hosts = get_hosts_from_describe_instances(describe_instances)
+        inventory = get_inventory_from_file(args.env)
+        update_inventory_public_ips(inventory, new_hosts)
+        save_inventory(args.env, inventory)
+
+
+STANDARD_ARGS = [
+    Provision,
+    Show,
+    Reip,
+]
```

Comment:
This is a cool pattern

Reply:
Thanks :) I started doing this in `commcare-cloud` and then copied that pattern here.